### PR TITLE
PYTHON-3601 OIDC: Add Documentation Examples

### PR DESCRIPTION
**Pull request by blink1073**

### CODE REVIEW

File: doc/examples/authentication.rst [hunk 1]
```diff
@@ -384,3 +384,165 @@ would be::
...
 .. _environment variables: https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html#configuration-envvars-runtime
+
+MONGODB-OIDC
+------------
+.. versionadded:: 4.7
+
+The `MONGODB-OIDC authentication mechanism`_ is available in MongoDB 7.0+ on Linux platforms.
+
+The MONGODB-OIDC mechanism authenticates using an OpenID Connect (OIDC) access token.
+The driver supports OIDC for  workload identity, defined as an identity you assign to a software workload
```
Comment (NoahStapp) [inline doc/examples/authentication.rst:395]:
```suggestion
The driver supports OIDC for workload identity, defined as an identity you assign to a software workload
```

File: doc/examples/authentication.rst [hunk 2]
```diff
@@ -384,3 +384,165 @@ would be::
...
+you can use the built-in support for Azure, where "<client_id>" below is the client id of the Azure
+managed identity, and ``<audience>`` is the url-encoded ``audience`` `configured on your MongoDB deployment`_.
+
+.. code-block:: python
+
+    import os
+
+    uri = os.environ["MONGODB_URI"]
+
+    props = dict(ENVIRONMENT="azure", TOKEN_RESOURCE="<audience>")
```
Comment (NoahStapp) [inline doc/examples/authentication.rst:420]:
```suggestion
    props = {"ENVIRONMENT": "azure", "TOKEN_RESOURCE": "<audience>"}
```

Nitpicky, but we should probably steer away from using `dict` in examples for readability/performance reasons.

File: doc/examples/authentication.rst [hunk 3]
```diff
@@ -384,3 +384,165 @@ would be::
...
+you can use the built-in support for GCP, where ``<audience>`` below is the url-encoded ``audience``
+`configured on your MongoDB deployment`_.
+
+.. code-block:: python
+
+    import os
+
+    uri = os.environ["MONGODB_URI"]
+
+    props = dict(ENVIRONMENT="gcp", TOKEN_RESOURCE="<audience>")
```
Comment (NoahStapp) [inline doc/examples/authentication.rst:446]:
```suggestion
    props = {"ENVIRONMENT": "gcp", "TOKEN_RESOURCE": "<audience>")
```

File: doc/examples/authentication.rst [hunk 4]
```diff
@@ -384,3 +384,165 @@ would be::
...
+
+    class MyCallback(OIDCCallback):
+        def fetch(self, context: OIDCCallbackContext) -> OIDCCallbackResult:
+            with open(os.environ["AWS_WEB_IDENTITY_TOKEN_FILE"]) as fid:
+                token = fid.read()
+            return OIDCCallbackResult(access_token=token)
+
+
+    uri = os.environ["MONGODB_URI"]
+    props = dict(OIDC_CALLBACK=MyCallback())
```
Comment (NoahStapp) [inline doc/examples/authentication.rst:478]:
```suggestion
    props = {"OIDC_CALLBACK": MyCallback()}
```

File: doc/examples/authentication.rst [hunk 5]
```diff
@@ -384,3 +384,165 @@ would be::
...
+
+
+  class MyCallback(OIDCCallback):
+      def fetch(self, context: OIDCCallbackContext) -> OIDCCallbackResult:
+          credential = DefaultAzureCredential(managed_identity_client_id=client_id)
+          token = credential.get_token(f"{audience}/.default").token
+          return OIDCCallbackResult(access_token=token)
+
+
+  props = dict(OIDC_CALLBACK=MyCallback())
```
Comment (NoahStapp) [inline doc/examples/authentication.rst:512]:
```suggestion
  props = {"OIDC_CALLBACK": MyCallback()}
```

File: doc/examples/authentication.rst [hunk 6]
```diff
@@ -384,3 +384,165 @@ would be::
...
+
+    class MyCallback(OIDCCallback):
+        def fetch(self, context: OIDCCallbackContext) -> OIDCCallbackResult:
+            with open("/var/run/secrets/kubernetes.io/serviceaccount/token") as fid:
+                token = fid.read()
+            return OIDCCallbackResult(access_token=token)
+
+
+    uri = os.environ["MONGODB_URI"]
+    props = dict(OIDC_CALLBACK=MyCallback())
```
Comment (NoahStapp) [inline doc/examples/authentication.rst:537]:
```suggestion
  props = {"OIDC_CALLBACK": MyCallback()}
```

**Review by NoahStapp — APPROVED** [review]
Minor nitpicks, otherwise looks great!

**Review by NoahStapp — CHANGES_REQUESTED** [review]
Minor nitpicks, otherwise looks great!

**Comment by blink1073** [comment]
ping @NoahStapp for final review.  :)

**Review by NoahStapp — APPROVED** [review]

**Pull request merged**